### PR TITLE
Add AspNetCore

### PR DESCRIPTION
**Pull request by crummel**

### CODE REVIEW

**Comment by crummel** [comment]
@JunTaoLuo, it looks like some projects aren't using the package version variables and are picking up outside versions.  Is this something you could take a look at?

```
 /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview7.19352.8 to 3.0.0-preview6.19264.9. Reference the package directly from the project to select a different version.  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.AspNetCore.DataProtection -> Microsoft.Win32.Registry 4.6.0-preview7.19352.8 -> System.Security.AccessControl 4.6.0-preview7.19352.8 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview7.19352.8)  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.NETCore.Platforms (>= 3.0.0-preview6.19264.9) [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview7.19352.8 to 3.0.0-preview6.19264.9. Reference the package directly from the project to select a different version.  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.AspNetCore.DataProtection -> System.Security.Cryptography.Xml 4.6.0-preview7.19352.8 -> System.Security.Permissions 4.6.0-preview7.19352.8 -> System.Security.AccessControl 4.6.0-preview7.19352.8 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview7.19352.8)  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/ref/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.NETCore.Platforms (>= 3.0.0-preview6.19264.9) [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview7.19352.8 to 3.0.0-preview6.19264.9. Reference the package directly from the project to select a different version.  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.AspNetCore.DataProtection -> Microsoft.Win32.Registry 4.6.0-preview7.19352.8 -> System.Security.AccessControl 4.6.0-preview7.19352.8 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview7.19352.8)  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.NETCore.Platforms (>= 3.0.0-preview6.19264.9) [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview7.19352.8 to 3.0.0-preview6.19264.9. Reference the package directly from the project to select a different version.  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.AspNetCore.DataProtection -> System.Security.Cryptography.Xml 4.6.0-preview7.19352.8 -> System.Security.Permissions 4.6.0-preview7.19352.8 -> System.Security.AccessControl 4.6.0-preview7.19352.8 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview7.19352.8)  [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
  /src/bin/src/AspNetCore.503ef7d96a55b6699094742ed4fc5792cf4e9d85/src/Antiforgery/src/Microsoft.AspNetCore.Antiforgery.csproj : error NU1605:  Microsoft.AspNetCore.Antiforgery -> Microsoft.NETCore.Platforms (>= 3.0.0-preview6.19264.9) [/src/Tools/dotnetcli3/sdk/3.0.100-preview5-011568/NuGet.targets]
```
(and more in CI)

**Comment by JunTaoLuo** [comment]
I'll take a look.

**Comment by JunTaoLuo** [comment]
Hmm I think there's a mix of preview6 and preview7 builds in here. I see that the preview6 version of Microsoft.NETCore.Platforms being compiled (and probably used implicitly by the SDK). That would explain the lines `Microsoft.AspNetCore.Antiforgery -> Microsoft.NETCore.Platforms (>= 3.0.0-preview6.19264.9)`. However, I see that the version of AspNetCore that you are using is https://github.com/aspnet/AspNetCore/commit/503ef7d96a55b6699094742ed4fc5792cf4e9d85 which is a preview 7 commit. Is this expected? If so, we'll need to override the version specified in https://github.com/aspnet/AspNetCore/blob/503ef7d96a55b6699094742ed4fc5792cf4e9d85/eng/Versions.props#L62-L83.

I was under the impression we were trying to build "coherent" sets of commits?

**Comment by crummel** [comment]
@omajid, while bringing this in I found that AspNetCore doesn't currently support a non-portable core runtime.  In the Microsoft official build, the portable core runtime and portable ASP.NET runtime are actually used even in the non-portable SDK.  I think matching the official build would be a good thing for us to do here, because otherwise source-build users are using bits that aren't really tested or used anywhere else in the same configuration, and it would be a non-trivial amount of work for ASP.NET to support essentially a new feature only used by source-build.  My proposal is to match the official build and continue building a non-portable SDK, but use the portable runtime with it.  Does that seem reasonable?

cc @leecow

**Comment by crummel** [comment]
Changes made:
- ASP.NET repos made coherent.
- Remove CLI and higher repos in the stack skipping ASP.NET.
- Fix repo references.
- Convert to always use portable core runtime.
- Add ASP.NET dev feed to smoke-test and re-enable most web tests.
- Output package versions as `{name}Version` in addition to `{name}PackageVersion`.  We used to do this in RestoreSources.props but some repos don't use that file anymore.

Remaining issues:
- Non-empty F# web projects are not currently working.  @JunTaoLuo says these requires bits not shipped in the SDK.  I'm looking into why we don't pull these bits at build-time like the official build does.
- Need to remove a bunch of test prebuilts in ASP.NET core - looking for a way to do these all at once.
- Prebuilt updates in general.
- Upstreaming patches to ASP.NET.

**Comment by crummel** [comment]
cc @dagood for core-setup changes if you'd like to take a look.

**Comment by omajid** [comment]
> My proposal is to match the official build and continue building a non-portable SDK, but use the portable runtime with it. Does that seem reasonable?

Doesn't seem like we have too many other options :(

Just so I understand the change, is it correct to summarize this as: .NET Core runtime will be portable, the ASP.NET Core runtime will be portable and the .NET Core SDK will be non-portable?

**Comment by crummel** [comment]
> .NET Core runtime will be portable, the ASP.NET Core runtime will be portable and the .NET Core SDK will be non-portable?

Yes, that's correct.

File: repos/core-setup.proj [hunk 1]
```diff
@@ -3,21 +3,13 @@
...
-
-    <BuildArguments>$(FlagParameterPrefix)restore $(FlagParameterPrefix)build</BuildArguments>
-    <BuildArguments>$(BuildArguments) $(FlagParameterPrefix)configuration $(Configuration)</BuildArguments>
-    <BuildArguments>$(BuildArguments) $(FlagParameterPrefix)ci</BuildArguments>
-    <BuildArguments>$(BuildArguments) /p:PortableBuild=$(OverridePortable)</BuildArguments>
-    <BuildArguments Condition="$(Platform.Contains('arm'))">$(BuildArguments) /p:TargetArchitecture=$(Platform) /p:DisableCrossgen=true /p:CrossBuild=true</BuildArguments>
+    <BuildArguments>--restore --build</BuildArguments>
+    <BuildArguments>$(BuildArguments) --configuration $(Configuration)</BuildArguments>
+    <BuildArguments>$(BuildArguments) --ci</BuildArguments>
+    <BuildArguments Condition="$(Platform.Contains('arm'))">$(BuildArguments) -TargetArchitecture=$(Platform) -DisableCrossgen=true -CrossBuild=true</BuildArguments>
```
Comment (dagood) [inline repos/core-setup.proj]: `/p:` => `-` accidental regression due to merge conflict? (Not that we exercise it... but nice to keep it alive.)

File: smoke-test.sh [hunk 2]
```diff
@@ -266,8 +259,11 @@ function runWebTests() {
     doCommand C# razor "$@" new restore run
 
     doCommand F# web "$@" new restore run
-    doCommand F# mvc "$@" new restore run
-    doCommand F# webapi "$@" new restore run
+    # XXX workaround for F# web template issue
+    # non-empty F# web tests are not currently expected to pass because they rely on code that does not ship in the SDK
+    # doCommand F# mvc "$@" new restore run
+    # doCommand F# webapi "$@" new restore run
+    # XXX end workaround
```
Comment (dagood) [inline smoke-test.sh]: What does a commented out workaround section mean?

Comment (crummel) [reply smoke-test.sh]: Hopefully made the intent here more clear; we now emit a warning that we're skipping known-failing tests with a link to the issue.

File: patches/aspnetcore/0008-Add-missing-reference.patch [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+diff --git a/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj b/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj
+index 52711dccf0..960e572814 100644
+--- a/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj
++++ b/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj
+@@ -14,6 +14,7 @@
+     <Reference Include="Microsoft.AspNetCore.Http.Features" />
+     <Reference Include="Microsoft.Extensions.ActivatorUtilities.Sources" />
+     <Reference Include="System.IO.Pipelines" />
++    <Reference Include="System.Threading.Tasks.Extensions" />
```
Comment (JunTaoLuo) [inline patches/aspnetcore/0008-Add-missing-reference.patch]: Hmm This seems suspicious. Why is this needed?

File: repos/aspnetcore.proj [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    <ShippingPackagesOutput>$(ProjectDirectory)artifacts/packages/$(Configuration)/Shipping/</ShippingPackagesOutput>
+    <NonShippingPackagesOutput>$(ProjectDirectory)artifacts/packages/$(Configuration)/NonShipping/</NonShippingPackagesOutput>
+    <RepoApiImplemented>false</RepoApiImplemented>
+    <DependencyVersionInputRepoApiImplemented>true</DependencyVersionInputRepoApiImplemented>
+
+    <GlobalJsonFile>$(ProjectDirectory)global.json</GlobalJsonFile>
+    <NuGetConfigFile>$(ProjectDirectory)NuGet.config</NuGetConfigFile>
+
+    <EnvironmentRestoreSources>$(SourceBuiltPackagesPath)</EnvironmentRestoreSources>
+    <EnvironmentRestoreSources Condition="'$(OfflineBuild)' == 'true'">$(EnvironmentRestoreSources)%3B$(ReferencePackagesDir)%3B$(PrebuiltPackagesPath)</EnvironmentRestoreSources>
```
Comment (dseefeld) [inline repos/aspnetcore.proj]: Should this include reference-packages as well?

File: smoke-test.sh [hunk 5]
```diff
@@ -259,17 +253,6 @@ function runAllTests() {
...
-    doCommand C# mvc "$@" new restore run
-    doCommand C# webapi "$@" new restore run
-    doCommand C# razor "$@" new restore run
-
-    doCommand F# web "$@" new restore run
-    doCommand F# mvc "$@" new restore run
-    doCommand F# webapi "$@" new restore run
-}
-
 function resetCaches() {
```
Comment (dseefeld) [inline smoke-test.sh]: Are we not running web tests anymore?  If you're removing this function, you should remove calls to it as well.

File: patches/aspnetcore/0008-Add-missing-reference.patch (hunk 3 above)
Comment (crummel) [reply patches/aspnetcore/0008-Add-missing-reference.patch]: Yeah, I tried this because of this error:
` ConnectionContext.cs(18,47): error CS0012: The type 'ValueTask' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Threading.Tasks.Extensions, Version=4.2.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. [/home/chris/tarball/src/AspNetCore.4a42afc5aea63750638e118560d43db04bd9ccc2/src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.     csproj]`

but adding it actually caused the production build to fail.  Any idea what's going on here?

On source-build, we've committed to minimal patches, but as long as they're in-flight and will make it into the next release, we'll be including those as a separate "temporary patch" statistic, so no worries as long as we can get the ball rolling.

Comment (crummel) [reply patches/aspnetcore/0008-Add-missing-reference.patch]: Actually this is almost definitely caused by DanS's comment below.  Trying it out.

File: smoke-test.sh (hunk 5 above)
Comment (crummel) [reply smoke-test.sh]: We actually a duplicate function, this just removes one of them.

File: repos/aspnetcore.proj (hunk 4 above)
Comment (crummel) [reply repos/aspnetcore.proj]: That should be `ReferencePackagesDir` right?

Comment (dseefeld) [reply repos/aspnetcore.proj]: Yes

Comment (dseefeld) [reply repos/aspnetcore.proj]: It should include prebuilts and reference-packages.

Comment (crummel) [reply repos/aspnetcore.proj]: So are you saying for all cases, not just the offline one?

Comment (dseefeld) [reply repos/aspnetcore.proj]: No, for offline cases.  It needs to include 1) source-built, 2) reference-packages and 3) prebuilts.

Comment (crummel) [reply repos/aspnetcore.proj]: Ok, I think this covers that then.  We always include `SourceBuiltPackagesPath`, for offline we add `ReferencePackagesDir` and `PrebuiltPackagesPath`, and for online we add a bunch of NuGet/MyGet/blob feeds.

Comment (dseefeld) [reply repos/aspnetcore.proj]: Oh, yeah.  My mistake.  I was missing that ref-pkgs and prebuilt pkgs were on the same line.  Sorry.

Comment (crummel) [reply repos/aspnetcore.proj]: No problem!

File: patches/aspnetcore/0008-Add-missing-reference.patch (hunk 3 above)
Comment (crummel) [reply patches/aspnetcore/0008-Add-missing-reference.patch]: That didn't seem to fix the issue.  @dseefeld, do you know if there's any issues with our reference version of System.Threading.Tasks.Extensions right now?  I can see it's being included as a reference, the versions match, and it does include `ValueTask`.

Comment (dseefeld) [reply patches/aspnetcore/0008-Add-missing-reference.patch]: @crummel  I don't know of any issues.  You could remove just that reference-package from the tarball reference-packages/packages directory to see if that narrows it down to a problem with that package.

Comment (crummel) [reply patches/aspnetcore/0008-Add-missing-reference.patch]: Our ref version of this package apparently has a public key problem, removing it as a workaround for now and logged https://github.com/dotnet/source-build/issues/1186.

**Comment by JunTaoLuo** [comment]
Yay!
